### PR TITLE
Improve packaging [PEP 517 + 621]

### LINKED ISSUE #4193: Migration to pyproject.toml
The packaging of pylint is a little dated, old files are kept and do not serve any purpose anymore (debian packaging is done elsewhere). The information are in multiple places and we still have a compatibility with distutils when it has been replaced by setuptools a long time ago. This might benefit from a little poetry 😉 

### Additional context
https://github.com/PyCQA/pylint/issues/3686#issuecomment-791688280

### PR DESCRIPTION
**Pull request by cdce8p**
## Description
* Use isolated build environments [PEP 517](https://peps.python.org/pep-0517/) with `python -m build` instead of deprecated `python setup.py sdist bdist_wheel`
* Move project metadata from `setup.cfg` to `pyproject.toml` [PEP 621](https://peps.python.org/pep-0621/)
* Cleanup and remove `MANIFEST.in` -> entries haven't been used.
* ~Add `README.rst` to `MANIFEST.in`. Required to be able to build the `wheel` from `sdist` and still include the readme text in the project metadata.~
* Remove `setup.py`. With recent pip versions (`>= 21.0`), no longer required for editable installs. `setup.cfg` is enough.

**Metadata diff**
```diff
-Author: Python Code Quality Authority
-Author-email: code-quality@python.org
+Author-email: Python Code Quality Authority <code-quality@python.org>
...
-Keywords: static code analysis linter python lint
+Keywords: static code analysis,linter,python,lint
```

Closes #4193

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -1,13 +1 @@
...
-prune examples
-prune tests
-prune script
-exclude .*
-exclude Dockerfile
-exclude README.rst
-exclude pylintrc
-exclude requirements_*.txt
-exclude tox.ini
+include README.rst
```
Comment (Pierre-Sassoulas) [inline MANIFEST.in:1]: We can probably remove this entirely now ? Doesn't the pyproject.toml handles it ?

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you ! I wanted to wait for pep660 but what's left of the setup.cfg is really manageable in the end. This would close #4193

**Comment by Pierre-Sassoulas** [comment]
Is there no change in the contributor doc at all ?

**Comment by cdce8p** [comment]
> We can probably remove this entirely now ? Doesn't the pyproject.toml handles it ?

I first checked with just ´[build-system]` when I got the error. You're correct though, with the `[project]` table, it works without `MANIFEST.in`.

> Is there no change in the contributor doc at all ?

`pip install .` and `pip install -e .` continue to function as expected. The main change is in the way the release package is build. `python -m build` instead of `python setup.py sdist bdist_wheel` but that is handled by Github actions anyway.

**Comment by cdce8p** [comment]
**Update**

Everything took a bit longer then I expected, but it's ready now.
I've also opened https://github.com/PyCQA/astroid/pull/1670 to update `astroid`, too.

> `README.rst` and `MANIFEST.in`

Turns out if you remove the `MANIFEST.in` completely, `README.rst` will be included automatically. That's why the `wheel` build didn't fail. I've added `README.rst` explicitly now, just in case we do add some other files there in the future.

> Let's not remove setup.py just yet:
> https://github.com/dependabot/dependabot-core/issues/4483
>
> This is quite an annoying bug.

Thanks @DanielNoord! Didn't know about that. I've restored `setup.py` and added a comment.

**Review by Pierre-Sassoulas — APPROVED** [review]
Perfect ! We've come a long way from the 2000's packaging !

File: setup.cfg [hunk 2]
```diff
@@ -93,6 +28,11 @@ known_third_party = platformdirs, astroid, sphinx, isort, pytest, mccabe, six, t
 skip_glob = tests/functional/**,tests/input/**,tests/extensions/data/**,tests/regrtest_data/**,tests/data/**,astroid/**,venv/**
 src_paths = pylint
 
+[flake8]
```
Comment (DanielNoord) [inline setup.cfg:31]: Can't this go in `.flake8`?

**Review by DanielNoord — DISMISSED** [review]

File: setup.cfg (hunk 2 above)
Comment (cdce8p) [reply setup.cfg:31]: Let's do it the other way round and move the whole flake8 config into `setup.cfg`. In the past, we have also removed `.isort`. At some point we can move most / all of these configs into `pyproject.toml`.

Comment (Pierre-Sassoulas) [reply setup.cfg:31]: Strong agree. I don't think ``.flake8`` support pyproject.toml though right now though . But there's other tool we could move already, tox and black probably.

Comment (Pierre-Sassoulas) [reply setup.cfg:31]: (https://github.com/PyCQA/flake8/issues/234#issuecomment-812800832)

Comment (cdce8p) [reply setup.cfg:31]: > Strong agree. I don't think `.flake8` support pyproject.toml right now though . But there's other tool we could move already, tox and black probably.

I would recommend to keep `tox` (and `pylint`) separate. Both have quite complex configurations. It's much easier to have them in their individual files. We could however move `isort`, `pytest`, and `mypy`.

**Comment by DanielNoord** [comment]
Are you sure? These are the logs/statuses for `pydocstringformatter`, which also doesn't have a `setup.py`.
https://github.com/DanielNoord/pydocstringformatter

Schermafbeelding 2022-06-28 om 16 57 27
Schermafbeelding 2022-06-28 om 16 57 35

**Comment by cdce8p** [comment]
> Are you sure? These are the logs/statuses for `pydocstringformatter`, which also doesn't have a `setup.py`. https://github.com/DanielNoord/pydocstringformatter

Yeah! This is from `python-typing-updater`. Dependabot just opened a new PR for black with the last run a few hours ago. Not exactly sure, why it's not working for you though. Wild guess, maybe it's because you don't have a `project` table in your `pyproject.toml`? https://github.com/cdce8p/python-typing-update
Screen Shot 2022-06-28 at 18 30 49

My suggestion would be to just try it without `setup.py`. If it fails, I can open a PR to add it back.

**Comment by DanielNoord** [comment]
Let's do that. I think @Pierre-Sassoulas csn report whether the check is running successfully afterwards!

**Comment by cdce8p** [comment]
Seems like you were right @DanielNoord 😞 Triggered the run manually just now.
I'll open a PR shortly to add `setup.py` back.

Screen Shot 2022-06-28 at 23 42 21

That just raises the question: Why does it work in my repo https://github.com/cdce8p/python-typing-update?
A difference I could think of is the `requirements.txt` file. Both `pylint` and `pydocstringformatter` do only have `requirements*<...>.txt` files.

**Comment by DanielNoord** [comment]
@cdce8p I have no idea really... The annoying thing is that in de dependabot issue there is already a solution. But my complete lack of ruby experience makes it impossible to contribute any test so I hesitated to create a PR...

**Comment by cdce8p** [comment]
Followup PR: #7082

**Comment by DanielNoord** [comment]
Followup PR: #7249.

**Pull request merged**